agents-api/agents_api/embed_models_registry.py

```diff
@@ -0,0 +1,124 @@
+import numpy as np
+from typing import TypedDict, Any
+from dataclasses import dataclass
+from transformers import PreTrainedTokenizer
+from agents_api.clients.model import openai_client
+from agents_api.clients.embed import embed
+
+
+def normalize_l2(x):
+    x = np.array(x)
+    if x.ndim == 1:
+        norm = np.linalg.norm(x)
+        if norm == 0:
+            return x
+        return x / norm
+    else:
+        norm = np.linalg.norm(x, 2, axis=1, keepdims=True)
+        return np.where(norm == 0, x, x / norm)
+
+
+class ModelNotSupportedError(Exception):
+    def __init__(self, model_name):
+        super().__init__(f"model {model_name} is not supporrted")
+
+
+class EmbeddingInput(TypedDict):
+    instruction: str | None
+    text: str
+
+
+@dataclass
+class EmbeddingModel:
+    embedding_provider: str
+    embedding_model_name: str
+    original_embedding_dimensions: int
+    output_embedding_dimensions: int
+    context_window: int
+    tokenizer: Any
+
+    @classmethod
+    def from_model_name(cls, model_name: str):
+        try:
+            return _embedding_model_registry[model_name]
+        except KeyError:
+            raise ModelNotSupportedError(model_name)
+
+    def preprocess(self, inputs: list[EmbeddingInput]):
+        """Maybe use this function from embed() to truncate (if needed) or raise an error"""
+        pass
+
+    async def embed(
+        self, inputs: list[EmbeddingInput]
+    ) -> list[np.NDArray | list[float]]:
+        embeddings: list[np.NDArray | list[float]] = []
+        input = [f"{input.get('instruction', '')} {input['text']}" for input in inputs]
+
+        if self.embedding_provider == "julep":
+            embeddings = await embed(input)
+        elif self.embedding_provider == "openai":
+            embeddings = (
+                await openai_client.embeddings.create(
+                    input=input, model=self.embedding_model_name
+                )
+                .data[0]
+                .embedding
+            )
+
+        return self.normalize(embeddings)
+
+    def normalize(
+        self, embeddings: list[np.NDArray | list[float]]
+    ) -> list[np.NDArray | list[float]]:
+        return [
+            (
+                e
+                if len(e) <= self.original_embedding_dimensions
+                else normalize_l2(e[: self.original_embedding_dimensions])
+            )
+            for e in embeddings
+        ]
+
+
+_embedding_model_registry = {
+    "text-embeddins-3-small": EmbeddingModel(
+        embedding_provider="openai",
+        embedding_model_name="text-embeddins-3-small",
+        original_embedding_dimensions=1024,
+        output_embedding_dimensions=1024,
+        context_window=8192,
+        tokenizer=PreTrainedTokenizer.from_pretrained("text-embeddins-3-small"),
```

Comment:
`embeddins` misspelled everywhere here, also isnt the right tokenizer `tiktoken` for openai?